src/modules/stripe/stripe.service.ts

```diff
@@ -0,0 +1,169 @@
+      .map((image) => image.url);
+
+    const sessionMetadata: SessionMetadata = {
+      bookingId: booking.id,
+      userId: booking.user.id,
+    };
+    const session = await this.stripe.checkout.sessions.create({
+      line_items: [
+        {
+          price_data: {
+            currency: 'usd',
+            product_data: {
+              name: accommodation.name,
+              description: accommodation.description,
+              images: accommodationImage,
+            },
+            unit_amount: convertToCent(booking.totalPrice),
+          },
+          quantity: 1,
+        },
+      ],
+      mode: 'payment',
+      metadata: sessionMetadata,
+      // For now we redirect to the same page with query params
+      success_url: generateAfterPaymentUrl({
+        accommodationId: accommodation.id,
+        bookingId: booking.id,
+        clientURL: this.clientURL,
+        success: true,
+      }),
+      cancel_url: generateAfterPaymentUrl({
+        accommodationId: accommodation.id,
+        bookingId: booking.id,
+        clientURL: this.clientURL,
+        success: false,
+      }),
+    });
+
+    return { checkoutSessionUrl: session.url };
+  }
+
+  async handleWebhook(body: Buffer, signature: string) {
+    if (
+      !this.stripe.webhooks.signature.verifyHeader(
+        JSON.stringify(body),
+        signature,
+        this.webhookSecret,
+      )
+    ) {
+      throw new BadRequestException('Invalid webhook signature');
+    }
+
+    const event = this.constructEvent(body, signature);
+    if (!event) return { received: false };
+
+    switch (event.type) {
+      case 'payment_intent.succeeded':
+        return await this.handlePaymentIntent(event.data.object, 'success');
+      case 'payment_intent.payment_failed':
+        return await this.handlePaymentIntent(event.data.object, 'failed');
+      default:
+        throw new BadRequestException(`Unhandled event type ${event.type}`);
+    }
+  }
+
+  async handlePaymentIntent(
+    paymentIntent: Stripe.PaymentIntent,
+    paymentStatus: 'success' | 'failed',
+  ) {
+    const metadata = paymentIntent.metadata;
+    const bookingId = metadata.bookingId;
+    const userId = metadata.userId;
+
+    const booking = await this.bookingsService.getBooking(userId, bookingId);
+
+    if (!booking) {
+      throw new BadRequestException(
+        `Booking with id ${bookingId} does not exist`,
+      );
+    }
+
+    if (booking.user.id !== userId) {
+      throw new BadRequestException(
+        `Booking with id ${bookingId} does not belong to user with id ${userId}`,
+      );
+    }
+
+    if (booking.status != 'PENDING') {
+      throw new BadRequestException(
+        `Booking with id ${bookingId} is not in pending state`,
+      );
+    }
+
+    await this.bookingsService.updateStatus(
+      {
+        status:
+          paymentStatus === 'success'
+            ? BookingStatus.CONFIRMED
+            : BookingStatus.CANCELLED,
```

Comment:
does this mean that if payment fails booking will be cancelled ?

Reply:
can the user redo their payment after failure or do they have to create a new booking again?